KAFKA-2273: Sticky partition assignment strategy (KIP-54)

vahidhashemian opened the pull request:
This PR implements a new partition assignment strategy called "sticky", and it's purpose is to balance partitions across consumers in a way that minimizes moving partitions around, or, in other words, preserves existing partition assignments as much as possible.

This patch is co-authored with @rajinisivaram and @edoardocomar.

vahidhashemian committed:
KAFKA-2273: Sticky partition assignment strategy (KIP-54)

This PR implements a new partition assignment strategy called "sticky", as described in KIP-54, and it's purpose is to balance partitions across consumers in a way that minimizes moving partitions around, or in other words, preserving partition assignments as much as possible.

noslowerdna commented:
Does Kafka retain pre-fetched messages for partitions that remain assigned to the same consumer after the rebalance? If not, that offsets a lot of the benefit this could provide.

By the way, for a similar pull request (https://github.com/apache/kafka/pull/979) I was asked to create a Kafka Improvement Proposal document (https://cwiki.apache.org/confluence/display/KAFKA/KIP-49+-+Fair+Partition+Assignment+Strategy).

rajinisivaram commented:
@noslowerdna I believe Kafka consumers deliver pre-fetched messages if a partition is not reassigned. I imagine this would also be useful if other state is held in-memory by the consuming application.

vahidhashemian commented:
@gwenshap I was wondering if I should open a KIP for this feature. I noticed there is one for Fair Assignment Strategy (KIP-49), and this would fall under the same category. Just wanted to get your feedback (and confirmation on what @rajinisivaram mentioned above) before starting a KIP. Thanks.

gwenshap commented:
Since the community is on limited bandwidth toward the release, if I were you, I'd put a note to open a KIP and start a discussion immediately after the release.

There are many KIPs now, and quite a few are not getting the attention they deserve because everyone is so busy.

There is also a general question of whether we want new partition-assignors in Kafka itself, or should they be separate projects since they are pluggable.

vahidhashemian commented:
@gwenshap Thank you for your quick response. I understand the tight schedule and I certainly did not mean a KIP for the upcoming release. I'll start a KIP shortly after the release to get some feedback.

hachikuji commented:
@vahidhashemian apologies for the delay. I'll try to get to this soon.

allenxwang commented:
Does the current implementation have any dependency on Kafka 0.10 APIs or protocols? Can the same code work with Kafka 0.9 client and broker?

vahidhashemian commented:
@hachikuji No worries :)
@allenxwang There is no dependency to newer APIs or protocols, and this should work with the 0.9 release. The PR was initially submitted before the 0.10 release.

vahidhashemian committed:
Fixed the newly identified findBugs issues

vahidhashemian committed:
Improved stickiness for when there is uniform topic subscription across consumers

vahidhashemian committed:
Improved stickiness for when there is uniform topic subscription across consumers

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/StickyAssignor.java:
```diff
@@ -0,0 +1,591 @@
+/**
+ * The sticky assignor serves two purposes. First, it guarantees an assignment that is as balanced as possible, meaning either:
+ * - the numbers of topic partitions assigned to consumers differ by at most one; or
+ * - each consumer that has 2+ fewer topic partitions than some other consumer cannot get any of those topic partitions transfered to it.
```
typo: "transfered"

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/StickyAssignor.java:
```diff
@@ -0,0 +1,591 @@
+import org.slf4j.LoggerFactory;
+
+/**
+ * The sticky assignor serves two purposes. First, it guarantees an assignment that is as balanced as possible, meaning either:
```
Seems we're missing the second purpose.

vahidhashemian replied:
Thanks for catching. Will add it.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/StickyAssignor.java:
```diff
@@ -0,0 +1,591 @@
+ * Sticky (preserves 5 partition assignments):
+ * C1 [t1p0, t1p1, t0p0]
+ * C2 [t2p0, t2p1, t2p2]
+ *
```
Maybe it's worth adding a code snippet which shows how to use the sticky assignment with the rebalance listener? It's a little different than with "range" and "roundrobin" from memory.

vahidhashemian replied:
Sure, I'll try to add some snippet and borrow from the KIP.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/StickyAssignor.java:
```diff
@@ -0,0 +1,591 @@
+    /**
+     * Remove partition from partition assignments of the given consumer
+     */
+    private void deassignPartition(TopicPartition partition, String consumer, TreeSet<String> sortedCurrentSubscriptions, HashMap<TopicPartition, String> currentPartitionConsumer) {
```
nit: maybe "unassignPartition"?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/StickyAssignor.java:
```diff
@@ -0,0 +1,591 @@
+    private void deepCopy(Map<String, List<TopicPartition>> source, Map<String, List<TopicPartition>> dest) {
+        dest.clear();
+        for (Entry<String, List<TopicPartition>> entry: source.entrySet())
+            dest.put(entry.getKey(), new ArrayList<TopicPartition>(entry.getValue()));
```
nit: seems you can use `new ArrayList<>`

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/StickyAssignor.java:
```diff
@@ -0,0 +1,591 @@
+        return copy;
+    }
+
+    private HashMap<TopicPartition, String> deepCopy(HashMap<TopicPartition, String> source) {
```
You can replace this method with `new HashMap<>(source)`

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/StickyAssignor.java:
```diff
@@ -0,0 +1,591 @@
+            return ByteBuffer.wrap(bos.toByteArray());
+        } catch (Exception ioex) {
+            log.error("Failed to serialize currentAssignment", ioex);
+            return super.getUserData();
```
Might be better to just propagate the exception and fail. We wouldn't expect to actually hit this in practice.

vahidhashemian replied:
Wouldn't this require modifying the method signature in the super class too? I assume you meant adding a something like `... throws IOException`.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/StickyAssignor.java:
```diff
@@ -0,0 +1,591 @@
+            byte[] array = new byte[assignment.userData().remaining()];
+            assignment.userData().get(array);
+            ObjectInputStream ois = new ObjectInputStream(new ByteArrayInputStream(array));
+            currentAssignment = (Map<String, List<TopicPartition>>) ois.readObject();
```
Unless I'm misunderstanding something, it seems like we're giving the full group assignment to every member in the group. I expected instead that each member would only receive its own assignment for the current generation and that we would aggregate the individual assignments on the leader when we received the group subscriptions. If we send all the assignments, then the overall overhead grows quadratically with the number of members in the group.

vahidhashemian replied:
This piece was not part of the original PR. However, I later realized that in the case where a group leader dies none of the other members have access to the latest partition assignment of each member of the group; and stickiness would not be honored in that case. So, the decision was to send the assignments to all members to cover this scenario. This is mentioned in the KIP too. Your thoughts?

hachikuji replied:
The issue is just that it limits the scalability of the assignor. For 100 members, you need to write all 100 assignments 100 times. It makes the overhead of the rebalance quite heavy.

I'm not sure I get the point about the leader dying. Each member knows their own assignment and should include that as user data in the subscription when they join the group. Whoever becomes the new leader will be able to collect the assignments from the member subscription user data. Would that work?

vahidhashemian replied:
Thanks for this great idea. I think it should work. I'm just trying to adjust the PR to work this way.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/StickyAssignor.java:
```diff
@@ -0,0 +1,591 @@
+        try {
+            byte[] array = new byte[assignment.userData().remaining()];
+            assignment.userData().get(array);
+            ObjectInputStream ois = new ObjectInputStream(new ByteArrayInputStream(array));
```
It might be better to use a Kafkaesque schema definition.

vahidhashemian replied:
With your change of user data suggestion below (user data becoming a list of topic partitions) do you think we should still go for a schema definition?

hachikuji replied:
Yes, I think we ought to use a Kafka schema definition even for the user data so that we can avoid dependence on java-specific serializations.

vahidhashemian replied:
Sounds good. I'll work on that too. Thanks.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/StickyAssignor.java:
```diff
@@ -0,0 +1,591 @@
+     * @return true if all lists in the collection have the same members; false otherwise
+     */
+    private <T> boolean hasIdenticalListElements(Collection<List<T>> col) {
+        Set<Set<T>> set = new HashSet<Set<T>>(col.size());
```
Couldn't we could just iterate through the collection and ensure that each list equals the previous one.

vahidhashemian replied:
We could. Do you mean using the `List`'s `containsAll(...)`?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractPartitionAssignor.java:
```diff
@@ -71,13 +72,17 @@ public Subscription subscription(Set<String> topics) {
     }
 
+    protected ByteBuffer getUserData() {
+        return ByteBuffer.wrap(new byte[0]);
```
This could probably just be `null`.

hachikuji reviewed (commented):
Thanks for the patch! I left a few initial comments. I haven't fully grokked the assignment algorithm, but it would be good to get clarification on the assignment serialization which is not quite what I expected.

vahidhashemian committed:
Modify the approach to preserve assignments in case of a leader change

Instead of having all group members keep a copy of all member assignments, leverage the `userData` field in `Subscription` object of each member, and have each member announce their previous assignment to the leader prior to leader performing the assignment.

vahidhashemian commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerProtocol.java:
```diff
@@ -78,6 +78,8 @@
             new Field(TOPIC_PARTITIONS_KEY_NAME, new ArrayOf(TOPIC_ASSIGNMENT_V0)),
             new Field(USER_DATA_KEY_NAME, Type.NULLABLE_BYTES));
+    public static final Schema TOPIC_PARTITION_ASSIGNMENT_V0 = new Schema(
+            new Field(TOPIC_PARTITIONS_KEY_NAME, new ArrayOf(TOPIC_ASSIGNMENT_V0)));
```
This is my initial attempt to use a schema for member assignments that are serialized/deserialized. Not so sure if this is the right place for defining it, or whether I needed to up the version. Your feedback is appreciated.

hachikuji replied:
Hmm.. I was expecting that the serialization would be custom to `StickyAssignor`. We have the UserData field specifically to let assignors embed their own metadata. So I would expect that we move the assignment schema into `StickyAssignor` and add logic to deserialize the UserData field when we receive the subscriptions. Does that make sense?

vahidhashemian replied:
Yes. that makes sense. I'll try to move these additions inside the `StickyAssignor` class. Thanks.

vahidhashemian reviewed (commented):
@hachikuji I tried to address your comments and feedback in the new commit. Thanks again for the suggestion to override the `subscription(...)` method to remove the redundancy in the original submission. Please take a look when you have a chance.

rogeralsing commented:
Is this going to get merged or is it being evaluated?
Having sticky partitions would be fantastic for stateful consumers/actor systems.
It wasnt completely clear from the Jira page if this is going in or not

vahidhashemian commented:
@rogeralsing Thanks for checking. This PR is currently under review. It's currently scheduled to be part of the next release (0.11.0.0).

vahidhashemian committed:
Fixed minor bugs and improved the stickiness algorithm

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerPair.java:
```diff
@@ -0,0 +1,113 @@
+import java.util.List;
+import java.util.Set;
+
+public class ConsumerPair {
```
Can you add some documentation for this class? Also, does it need to be public? Perhaps it can go in `consumer.internals`?

vahidhashemian replied:
I'll try to move both `ConsumerPair` and `PartitionMovements` inside the `StickyAssignor` class and limit the exposures as much as possible. Thanks.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/PartitionMovements.java:
```diff
@@ -0,0 +1,165 @@
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.common.utils.Utils;
+
+public class PartitionMovements {
```
Similarly, let's make sure we're not exposing anything that doesn't need it.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/StickyAssignor.java:
```diff
@@ -0,0 +1,706 @@
+            dest.put(entry.getKey(), new ArrayList<>(entry.getValue()));
+    }
+
+    protected Map<String, List<TopicPartition>> deepCopy(Map<String, List<TopicPartition>> assignment) {
```
nit: could this be private? A few more of these below. Since `StickyAssignor` is a public API, we need to be a little extra careful about what we expose. If it is exposed for testing, perhaps we can move it to a utility class in `consumer.internals`.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/StickyAssignor.java:
```diff
@@ -0,0 +1,706 @@
+    }
+
+    @SuppressWarnings("serial")
+    protected static class PartitionAssignmentComparator implements Comparator<String>, Serializable {
```
Does this really need to be `Serializable`? Same for the other comparators above.

vahidhashemian replied:
We should be okay without `Serializable`. Thanks.

vahidhashemian replied:
Just realized why I made these `Serializable`. It was because findbugs issues a warning:
>  SE_COMPARATOR_SHOULD_BE_SERIALIZABLE: Comparator doesn't implement Serializable
This class implements the Comparator interface. You should consider whether or not it should also implement the Serializable interface. If a comparator is used to construct an ordered collection such as a TreeMap, then the TreeMap will be serializable only if the comparator is also serializable. As most comparators have little or no state, making them serializable is generally easy and good defensive programming. 

What's the right approach here? Should we suppress them? Thanks.

hachikuji replied:
Wow, that's annoying. No harm making the Serializable I guess, but perhaps we can just add the `serialVersionUID` so that we don't need the suppressions.

hachikuji reviewed (commented):
Left a few more comments. I think the main thing is to make sure we are making use of the subscription UserData field to pass the previous assignments in the subscription data.

vahidhashemian committed:
Addressed feedback received on 05/16

- Moved `ConsumerPair` and `PartitionMovements` definitions inside `StickyAssignor`
- Moved the schema definition for topic partition list and its serializer/deserializer inside `StickyAssignor`
- Limited the exposure of newly added class members

vahidhashemian commented:
@hachikuji I tried to address your recent feedback in the latest commit. Please advise if you still see issues (specially with moving the schema definition inside the `StickyAssignor` class). Thanks.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/StickyAssignor.java:
```diff
@@ -0,0 +1,951 @@
+    public static final Schema TOPIC_PARTITION_ASSIGNMENT_V0 = new Schema(
+            new Field(ConsumerProtocol.TOPIC_PARTITIONS_KEY_NAME, new ArrayOf(ConsumerProtocol.TOPIC_ASSIGNMENT_V0)));
+
+    protected Map<String, List<TopicPartition>> currentAssignment = new HashMap<>();
```
Maybe make this package access?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/StickyAssignor.java:
```diff
@@ -0,0 +1,951 @@
+
+    // this schema is used for preserving consumer's assigned partitions list and
+    // sending it to the leader during a rebalance
+    public static final Schema TOPIC_PARTITION_ASSIGNMENT_V0 = new Schema(
```
Can this be private?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/StickyAssignor.java:
```diff
@@ -0,0 +1,951 @@
+    private List<TopicPartition> memberAssignment = null;
+    private PartitionMovements partitionMovements;
+
+    private void deepCopy(Map<String, List<TopicPartition>> source, Map<String, List<TopicPartition>> dest) {
```
nit: maybe utility methods can be moved to the bottom of the class.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/StickyAssignor.java:
```diff
@@ -0,0 +1,951 @@
+    // this schema is used for preserving consumer's assigned partitions list and
+    // sending it to the leader during a rebalance
+    public static final Schema TOPIC_PARTITION_ASSIGNMENT_V0 = new Schema(
+            new Field(ConsumerProtocol.TOPIC_PARTITIONS_KEY_NAME, new ArrayOf(ConsumerProtocol.TOPIC_ASSIGNMENT_V0)));
```
I realize that these values are the same that are used in the consumer protocol, but perhaps we should just copy the data so there is no unneeded dependence.

vahidhashemian replied:
Sure. I'll try to remove the dependency to `ConsumerProtocol` in the next commit.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/StickyAssignor.java:
```diff
@@ -0,0 +1,951 @@
+        struct.set(ConsumerProtocol.TOPIC_PARTITIONS_KEY_NAME, topicAssignments.toArray());
+        ByteBuffer buffer = ByteBuffer.allocate(ConsumerProtocol.CONSUMER_PROTOCOL_HEADER_V0.sizeOf() +
+                                                TOPIC_PARTITION_ASSIGNMENT_V0.sizeOf(struct));
+        ConsumerProtocol.CONSUMER_PROTOCOL_HEADER_V0.writeTo(buffer);
```
Is the goal to have a version number for this schema? If so, perhaps we could just add it to the sticky assignor schema. Something like this:

```java
    public static final Schema STICKY_ASSIGNOR_USER_DATA_V0 = new Schema(
            new Field("version", Type.INT16),
            new Field("previous_assignment", TOPIC_PARTITION_ASSIGNMENT_V0));
```
That would remove the dependence on the consumer protocol.

If we add a version, the next question is what kind of compatibility we are attempting to provide. Forward compatibility is probably the only useful thing to provide in this assignor. When you're doing a rolling upgrade to the newer assignor version, there's no way to ensure that only the newer member becomes a group leader, which means backwards compatibility is not enough. So if we can't have forward compatibility, then perhaps we can just leave the version out. To get it, we would basically need the parser to just ignore version numbers which are higher than it expects.

vahidhashemian replied:
I'll try to skip the version for now, as it may not bring a lot of value as you described. Hopefully, the next update makes it better :)

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/StickyAssignor.java:
```diff
@@ -0,0 +1,951 @@
+        }
+    }
+
+    protected static class SubscriptionComparator implements Comparator<String>, Serializable {
```
nit: these could all be private or package access?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractPartitionAssignor.java:
```diff
@@ -34,6 +34,7 @@
  */
 public abstract class AbstractPartitionAssignor implements PartitionAssignor {
     private static final Logger log = LoggerFactory.getLogger(AbstractPartitionAssignor.class);
+    private Map<String, Subscription> subscriptions = null;
```
Instead of creating a back channel to propagate this to extensions, maybe we should change the signature of the abstract `assign` method below:
```java
public abstract Map<String, List<TopicPartition>> assign(
  Map<String, Integer> partitionsPerTopic,
  Map<String, Subscription> subscriptions);
```
We can do this in a follow-up if you like.

vahidhashemian replied:
Wouldn't this change affect the existing assignors too?

hachikuji replied:
Yes, it would, but only their internals.

vahidhashemian committed:
Minor updates to address feedback
* Limited the scope of some classes and members
* Removed `StickyAssignor`'s dependency to `ConsumerProtocol`
* Simplified the 'previous_assignment' schema used in `StickyAssignor`

vahidhashemian commented:
@hachikuji Thanks for the feedback. For now, I tried to addressed them all except for the last one about modifying the signature of `assign`.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/StickyAssignor.java:
```diff
@@ -0,0 +1,947 @@
+
+        prepopulateCurrentAssignments();
+        // make a deep copy of currentAssignment
+        Map<String, List<TopicPartition>> oldAssignment = deepCopy(currentAssignment);
```
I think I'm not getting why we depend on `currentAssignment`. Wouldn't we expect all of the assignments to be received through the subscriptions?

hachikuji replied:
OK, I think I see. We populate `currentAssignment` based on the subscriptions. Is there actually any need to cache it after the assignment is made?

vahidhashemian replied:
`currentAssignment` is calculated through the subscription user data in the `prepopulateCurrentAssignments()` method a couple of lines above. I hope I understood the question correctly.

hachikuji replied:
Not sure if you saw my last comment. I'm wondering if it is necessary to keep the current assignment since we always repopulate it from the subscriptions.

vahidhashemian replied:
Just saw it. We don't have to keep it, it's just that almost all unit tests are written based on the assumption that we keep it (since we are not running any actual consumers in the unit tests and are just testing the assignment algorithm; so no subscription user data exists). I could look into modifying the unit tests and using consumers.

hachikuji replied:
If you want to save this for a follow-up, that is fine. But perhaps we should at least make sure we clear the current assignment after every rebalance?

vahidhashemian replied:
Yeah, it would be better to leave this for a follow-up, if that's okay.
For now, we can clear `currentAssignment` if there is subscription user data. Would that work as a temporary solution?

hachikuji replied:
Hmm.. why not always clear it? The behavior becomes a bit less predictable if it depends on state which is not part of the current rebalance.

vahidhashemian replied:
The issue is if we always clear it, for the unit tests that don't run any `KafkaConsumer`, there is no way to preserve previous assignment (the `subscription` user data is null). Perhaps I am missing something?

hachikuji replied:
But you can simulate a previous assignment by setting the previous assignment in the subscription, right? You don't actually need to have executed the previous assignment.

vahidhashemian replied:
Is this assuming that we already changed the `assign()` signature as you suggested before?

hachikuji replied:
OK. Let's leave it as it is for now. My larger point is that the sticky assignor should be a pure function. It takes inputs (subscriptions) and generates outputs (assignments). There shouldn't be any need to carry state between rebalances. It sounds like that is just a convenience for test purposes. In a follow-up, we should try to improve this.

vahidhashemian replied:
Sure, that sounds good. I agree with your point about it being a pure function.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/StickyAssignor.java:
```diff
@@ -0,0 +1,947 @@
+        return copy;
+    }
+
+    public static Map<String, List<Integer>> asMap(Collection<TopicPartition> partitions) {
```
This could be private?

hachikuji replied:
Should have thought of this before, but `CollectionUtils.groupDataByTopic` already does this.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/StickyAssignor.java:
```diff
@@ -0,0 +1,947 @@
+            superCycle.remove(superCycle.size() - 1);
+            superCycle.addAll(cycle);
+            Iterator<List<String>> it = cycles.iterator();
+            while (it.hasNext()) {
```
nit: this could be a normal "for each" loop.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/StickyAssignor.java:
```diff
@@ -0,0 +1,947 @@
+     * whether a partition reassignment results in cycles among the generated graph of consumer pairs.
+     */
+    private static class ConsumerPair {
+        private final String src;
```
Perhaps we could name these fields `srcMemberId` and `dstMemberId` to make it clear what they represent?

vahidhashemian committed:
Minor updates based on the feedback received on 5/17

hachikuji reviewed (approved):
LGTM. Thanks for the persistence on this KIP.
